File: workbench/mps-workbench/source/jetbrains/mps/intellij/integration/MPSRecentProjectsManagerBase.java
/*
 * Copyright 2000-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */
package jetbrains.mps.intellij.integration;

import com.intellij.ide.RecentProjectsManagerBase;
import com.intellij.ide.impl.OpenProjectTask;
import com.intellij.openapi.project.Project;
import kotlin.coroutines.Continuation;
import kotlin.jvm.functions.Function3;
import kotlinx.coroutines.CoroutineScope;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.nio.file.Path;

public class MPSRecentProjectsManagerBase extends RecentProjectsManagerBase {

  // TODO: 232 platform API change
  public MPSRecentProjectsManagerBase(@NotNull CoroutineScope coroutineScope) {
    super(coroutineScope);
  }

  @Nullable
  @Override
  public Object openProject(@NotNull Path projectFile, @NotNull OpenProjectTask options, @NotNull Continuation<? super Project> completion) {
    // TODO when kotlin compilation is available for idea runner, replace this by a kotlin class containing the openProject function from the kotlin file
    // No idea how to deal with suspend functions, we delegate this to the kotlin method
    try {
      final Class<?> kClass = Class.forName("jetbrains.mps.intellij.integration.MPSRecentProjectsManagerBaseKt");
      Method openProject = kClass.getMethod("openProject", Path.class, OpenProjectTask.class, Function3.class, Continuation.class);
      return openProject.invoke(null, projectFile, options, (Function3<Path, OpenProjectTask, Continuation<? super Project>, Object>) super::openProject, completion);
    } catch (InvocationTargetException | ClassNotFoundException | NoSuchMethodException | IllegalAccessException e) {
      throw new RuntimeException(e);
    }
  }
}

File: workbench/mps-workbench/source/jetbrains/mps/intellij/integration/MPSRecentProjectsManagerBase.kt
package jetbrains.mps.intellij.integration

import com.intellij.ide.impl.OpenProjectTask
import com.intellij.ide.impl.ProjectUtilCore
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.wm.impl.welcomeScreen.WelcomeFrame
import com.intellij.util.io.exists
import jetbrains.mps.workbench.actions.OpenMPSProjectTrustProjectHelperK
import java.nio.file.Path
import kotlin.io.path.absolutePathString

private val LOG = logger<MPSRecentProjectsManagerBase>()

suspend fun openProject(projectFile: Path?, options: OpenProjectTask?, superFun: (suspend (projectFile: Path, options: OpenProjectTask) -> Project?)?): Project? {
    //It seems that coroutines call this method by mistake with all null parameters values, if the method declares any of the parameters as mandatory.
    // This causes the problem of MPS-35253.
    if (projectFile == null) {
        LOG.error("MPSRecentProjectsManagerBase called with null path parameter", IllegalArgumentException("Unexpected null argument"))
        return null
    }
    if (options == null) {
        LOG.error("MPSRecentProjectsManagerBase called with null options parameter", IllegalArgumentException("Unexpected null argument"))
        return null
    }
    if (superFun == null) {
        LOG.error("MPSRecentProjectsManagerBase called with null superFun parameter", IllegalArgumentException("Unexpected null argument"))
        return null
    }

    if (!projectFile.exists() || !ProjectUtilCore.isValidProjectPath(projectFile)) {
        LOG.warn("MPSRecentProjectsManagerBase called with a non-existent project file: " + projectFile.absolutePathString())
        return null
    }

    val localOptions = options

    try {
        val trusted = OpenMPSProjectTrustProjectHelperK().checkTrust(projectFile)

        return if (trusted) {
            superFun(projectFile, localOptions)
        } else {
            LOG.info("Project is not trusted -> return null")
            null
        }
    } finally {
        WelcomeFrame.showIfNoProjectOpened()
    }
}